### PR TITLE
🌱 Add LogCollector to E2E framework

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR adds a `LogCollector` to the E2E framework.

The `LogCollector` is a pluggable component that accesses machines in a workload cluster and collect logs, and each provider can inject the specific implementation.

This PR includes the implementation for CAPD. 

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes-sigs/cluster-api/issues/3344

/assign @randomvariable

### CODE REVIEW

**Comment by ncdc** [comment]
Doesn't have to be in this PR, but I think we also need the pod logs from the workload cluster.

**Comment by fabriziopandini** [comment]
/hold
I would like to experiment a little bit on logs structure

File: test/framework/docker_logcollector.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+		),
+		execToPathFn(
+			"kubelet.log",
+			"journalctl", "--no-pager", "--output=short-precise", "-u", "kubelet.service",
+		),
+		execToPathFn(
+			"containerd-info.txt",
+			"crictl", "info",
+		),
+
```
Comment (vincepri) [inline test/framework/docker_logcollector.go:68]:
```suggestion
```

File: test/framework/docker_logcollector.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+			f, err := fileOnHost(filepath.Join(outputPath, outputFileName))
+			if err != nil {
+				return err
+			}
+			defer f.Close()
+			return execOnContainer(containerName, f, command, args...)
+		}
+	}
+	return errors.AggregateConcurrent([]func() error{
+		// record info about the node container
```
Comment (vincepri) [inline test/framework/docker_logcollector.go:47]: What do you mean here?

Comment (fabriziopandini) [reply test/framework/docker_logcollector.go:47]: What I was thinking about is that in CAPD the node == a docker container, but I can remove this line

Comment (vincepri) [reply test/framework/docker_logcollector.go:47]: Got it, yeah seems it's clear what we're collecting, up to you

**Comment by fabriziopandini** [comment]
Comment addressed + squashed commits
Now the logs structure is cleaner IMO, because all the clusters use the specName as a prefix and the self-hosted cluster now contains both the logs for the management cluster and the workload cluster
/hold cancel

**Comment by randomvariable** [comment]
Just had a browse through of the artifacts in the last e2e job

/lgtm
/assign @vincepri for approval and overrides.

**Comment by randomvariable** [comment]
/assign @vincepri 
for approval and override

**Review by vincepri — APPROVED** [review]
/override pull-cluster-api-test
/approve

**Pull request merged**